### ISSUE TITLE
🚀 Feature Request: Implement "list" command

### ISSUE THREAD

**Issue by alexandernst** [cloudflare/workers-sdk#3009]
### Describe the solution

It would be nice to be able to `list` files in a bucket, especially if listing could work with folder and wildcards.

Eg:

```
wrangler r2 object list my-bucket/*  <--- list all files
wrangler r2 object list my-bucket/*.jpg  <--- list all JPG files
wrangler r2 object list my-bucket/foo/* <-- list all files inside the "foo" folder
```

**Comment by JacobMGEvans** [comment]
Thank you for your suggestion, certainly would be a improvement to R2

**Comment by Gorniv** [comment]
This base command for work with bucket.
You lost 34 GB data in UI, I want to check is it exist by cli and can't, WTF?

**Comment by majal** [comment]
I agree that this a a really basic command that should be included. For now, here is my workaround in bash:

```
wrangler r2 object get ${bucket_name}/ | tr -d '\n'| sed 's/^[^{]\+//' | jq -r '.result[].key'
```

Even the result is not a valid JSON, so there is a need to manipulate the data first. And it truncates the results to the first 20 objects, `--help` does not show flags to disable truncate, or to get the next page.

Feels like R2 cli is still in experimental stage.

**Comment by ferhatelmas** [comment]
We gave a try to implement this #4070 but then team discussed internally and decided not duplicate various clients out there because wrangler is mainly to support script deployments.

For example, I would suggest using full functional client such as [mc](https://github.com/minio/mc):

```bash
> mc find my-account/my-bucket --name '*.jpg'
a.jpg
b.jpg
c.jpg
...
```

@admah could you close this issue?

**Comment by sdnts** [comment]
For a little more transparency here: the issue is that searching using a glob pattern isn't supported by R2 yet. This means that supporting this in Wrangler involves making multiple ListObjects calls to R2 (which only returns 1000 objects at a time), and resolving glob patterns on the client. Not only is this slow, but it can easily rack up a bill if you have lots of objects in your bucket, and this isn't immediately obvious.

We felt that manual scripting is generally a more controlled way to do this. Dedicated CLIs such as https://rclone.org also work, but they have the same issues.

Are y'all trying to verify uploads during development or is there a use case we haven't considered where you'd want this in wrangler?

**Comment by alexandernst** [comment]
Hi! Thanks for the detailed explanation.

We currently have a pipeline deploy process that checks what files exist in the bucket. There isn't anything we can't implement "manually" (as in "read top folders, perform some logic, walk subfolders, perform more logic, etc..."), but having actual filters would greatly reduce that complexity.

I understand your point about not having that capability in R2, and closing this ticket is totally reasonable. I'd rather not have the feature at all than having a half backed feature (fetching objects in batches of 1000 and filtering on the client side).

**Comment by ms-ati** [comment]
Given the statements above, is it supposed to be possible to connect `mc` or `rclone` to a locally-running `wrangler r2`? If so, how?

**Comment by bannsec** [comment]
yeah i'm running into this too. trying to test code locally with r2 interactions prior to deploying, but having a hard time actually seeing what's in my r2 store. this feels like a giant omission. do folks generally test their r2 setups in prod or something?

**Comment by dweekly** [comment]
Bummed wrangler doesn't support this directly. Listing files in a directory is a pretty elementary operation and pushing folks outside of the wrangler ecosystem to solve for this is a bummer.

Please implement `list`!

**Comment by Joelsz** [comment]
Is there any possible way to list all objects in a bucket? I'm using local wrangler for dev and there's no UI to check that.

**Comment by bannsec** [comment]
My solution has been to write a custom local worker that exposes the ability to list objects.

**Comment by mootari** [comment]
@Joelsz Assuming you only have one bucket the following should work:
```sh
sqlite3 --table .wrangler/state/v3/r2/miniflare-R2BucketObject/*.sqlite '
  select key, datetime(uploaded/1000, '\''unixepoch'\'') uploaded, http_metadata->'\''$.contentType'\'' contentType
  from _mf_objects order by 2'
```
Otherwise you'll have to specify enough of the hash to ensure that the globbing only matches one hash.

The hash itself can be calculated as follows:
```js
import {createHash, createHmac} from "node:crypto";

// Adapted from https://github.com/cloudflare/workers-sdk/blob/2df1d066c/packages/miniflare/src/plugins/shared/index.ts#L236-L249
function namespaced(uniqueKey: string, name: string) {
  const key = createHash("sha256").update(uniqueKey).digest();
  const nameHmac = createHmac("sha256", key).update(name).digest().subarray(0, 16);
  const hmac = createHmac("sha256", key).update(nameHmac).digest().subarray(0, 16);
  return Buffer.concat([nameHmac, hmac]).toString("hex");
}

console.log(namespaced("miniflare-R2BucketObject", THE_BUCKET_NAME));

**Comment by siracusa** [comment]
I, too, would like basic list functionality _for local development R2 buckets only_. 🙏
